### PR TITLE
Support NotNullIfNotNullAttribute

### PR DESCRIPTION
**Pull request by kerams**
Only supported on C# and F# methods. Functions use an entirely different pathway. Also only one instance of the attribute is supported because at that point we can't evaluate nullness of relevant arguments to link the right one (but maybe we could introduce a new type of nullness linkage). It's also exceedingly rare...

## Checklist

- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated

### CODE REVIEW

File: tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    |> typeCheckWithStrictNullness
+    |> shouldSucceed
+
+[<FactForNETCOREAPP>]
+let ``BCL Path.GetExtension - nullable input yields nullable result`` () =
+    FSharp """module MyLibrary
+open System.IO
+
+let maybeNull : string | null = "file.txt"
+let ext : string = Path.GetExtension(maybeNull)
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:84]: I remember especially the `Path` APIs being heavy users of that - can we eliminate some "useless" null checking now? ( I guess too much hassle to do it via ifdefs now?)

File: tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    cu
+    |> withStrictNullness
+    |> typecheck
+
+let csNotNullLib =
+    CSharp """
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+namespace NotNullLib {
+    public class C {
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:23]: Does C# support it also on generic arguments? (possibly also non-restricted to reference types, I remember some craziness with `T?` working )

Comment (kerams) [reply tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:23]: Yes,

```csharp
[return: NotNullIfNotNull(nameof(x))]
public static T? Echo<T>(T? x) => x;
```

File: tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs (hunk 1 above)
Comment (kerams) [reply tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:84]: Other than the 2 usages I had to adjust, it seems the return values are either used in a way where nullness doesn't matter (passing to `String.Equals`) or `Unchecked.nonNull` is used. I'd probably just strip the latter later once the new SDK has flowed back into the repo.

File: tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        [return: NotNullIfNotNull("input")]
+        public static string? Echo(string? input) => input;
+
+        // The result is non-null when the SECOND parameter is non-null.
+        [return: NotNullIfNotNull("second")]
+        public static string? DependsOnSecond(string? first, string? second) => second;
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:29]: Could you pls also do a test for type inference of an F# function argument ? And assert what is being inferred.

```
let f x =
     C.DependsOnSecond(x,x)
```

File: tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs (hunk 1 above)
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:84]: Can you pls also do one where the `null` literal is passed directly as a method argument? Both when inlined into the method application, or as vanilla `let x = null` above.

There have been some regressions due to how the literal is type inferred initially.

Comment (kerams) [reply tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:84]: Nice, this actually surfaced a bug where

```fsharp
let maybeNull = null
let ext : string = Path.GetExtension(maybeNull)
```

would pass, because `nullnessOfTy` on its own returns `WithoutNull` for `maybeNull`.

Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:84]: yep, makes sense to wait for it.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 4]
```diff
@@ -10505,6 +10559,27 @@ and TcMethodApplication
...
+                        | [ paramName ] ->
+                            match TryGetCallerArgType g baseMinfo callerArgs paramName with
+                            | Some callerArgTy ->
+                                let retTy = baseMinfo.GetFSharpReturnType(cenv.amap, mMethExpr, callerTyArgs)
+                                let argNullness =
+                                    match GetTyparTyIfSupportsNull g callerArgTy with
+                                    | ValueSome _ -> g.knownWithNull
+                                    | ValueNone -> nullnessOfTy g callerArgTy
+                                MethInfoWithModifiedReturnType(baseMinfo, replaceNullnessOfTy argNullness retTy)
+                            | None -> baseMinfo
```
Comment (T-Gro) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs:10645]: It is very degenerate, but it occured to me that Methods with C#-style optionals and/or ParamArray might end up in the None case here.

I am thinking if there isnt a higher-level concept already that would answer what TryGetCallerArgType does, but also follow the (complicated) ruleset of special method arguments.

Comment (kerams) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs:10645]: Added extra tests.

`CalledMeth` exposes that data, but it's constructed further down. I could either create temporary `CalledMeth` here or factor the bits we're interested in (code that constructs arg sets) out of the type, but that would be a fairly complicated change because of interdependencies.

Comment (T-Gro) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs:10645]: Yep, let's not do that.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 5]
```diff
@@ -10505,6 +10559,27 @@ and TcMethodApplication
...
+                        let baseMinfo = match minfo with MethInfoWithModifiedReturnType(inner, _) -> inner | _ -> minfo
+                        match NotNullIfNotNullParamNames g baseMinfo with
+                        | [ paramName ] ->
+                            match TryGetCallerArgType g baseMinfo callerArgs paramName with
+                            | Some callerArgTy ->
+                                let retTy = baseMinfo.GetFSharpReturnType(cenv.amap, mMethExpr, callerTyArgs)
+                                let argNullness =
+                                    match GetTyparTyIfSupportsNull g callerArgTy with
+                                    | ValueSome _ -> g.knownWithNull
+                                    | ValueNone -> nullnessOfTy g callerArgTy
```
Comment (T-Gro) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs]: I believe there are more edge cases, 'AllowNull, option-like and unit types come to mind.

Maybe replace this logical block with 'TypeNullIsExtraValueNew || TypeNullIsTrueValue' (the first is about null from the F# perspective, TypeNullIsTrueValue takes care about runtime representation for interop scenarios - think passing 'None' or '()' to a C# method).

File: tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs (hunk 2 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Language/Nullness/NotNullIfNotNullTests.fs:23]: @kerams :

What about adding 2 additional echo methods, a generic one and an `object` accepting one.
And doing parametrized tests for the various edge cases of F#<>runtime interop when it comes to nullness:
unit
None
type with AllowNull
`T | null` - passed from a generic function

File: src/Compiler/Checking/Expressions/CheckExpressions.fs (hunk 5 above)
Comment (kerams) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs]: `TypeNullIsExtraValueNew` calls `.Evaluate()`, forcing a solution too early, which causes `Csharp NotNullIfNotNull - unannotated parameter with non-null return annotation fails` to fail.

Comment (T-Gro) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs]: Hm I believe this might be a bug in its implementation, TryEvaluate() should do.

In general I would prefer if we used the TypedTree helpers to answer "can this be null in F#" and "can this be null in IL".

Comment (kerams) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs]: Never mind, the function wasn't the problem - what I did in the else branch was...

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 6]
```diff
@@ -3340,6 +3340,46 @@ let GetMethodArgs arg =
...
+    callerArgs.Named
+    |> List.tryPick (List.tryPick (fun (CallerNamedArg(id, arg)) -> if id.idText = paramName then Some arg else None))
+    |> Option.orElseWith (fun () ->
+        // If there is no matching named argument, find the argument in the same position as the parameter with the given name
+        minfo.GetParamNames()
+        |> Seq.concat
+        |> Seq.tryFindIndex (fun nm -> match nm with Some nm -> nm = paramName | _ -> false)
+        |> Option.bind (fun idx -> Seq.concat callerArgs.Unnamed |> Seq.tryItem idx)
+    )
+    |> Option.map (fun arg -> tyOfExpr g arg.Expr)
```
Comment (T-Gro) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs:3400]: 🤖🕵️

```fsharp
// C#: [return: NotNullIfNotNull("s")] static string? RefEcho(ref string? s) => s;
let mutable s : string | null = null
(C.RefEcho(&s)).Length   // no FS3261 -> NullReferenceException
```
`tyOfExpr g arg.Expr` on a `byref` arg is `byref<string|null>` (never null), so the return is typed non-null and the deref is unguarded. By-value `NotNullIfNotNull` and plain by-ref both still warn; `main` warns here too. Read the nullness of the byref **contents**.

**Review by T-Gro — APPROVED** [review]
The byref scenario is my last concern, otherwise LGTM 👍 . (@abonie can re-approve if there are new pushes)

**Comment by kerams** [comment]
I'm not sure auto-merge does anything, what with the broken release notes check:)

**Review by abonie — APPROVED** [review]
